### PR TITLE
Add analyzers/fixers for Environment.ProcessPath and CurrentManagedThreadId

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/43257
Fixes https://github.com/dotnet/runtime/issues/42948

cc: @carlossanlop, @mavasani, @jkotas

### CODE REVIEW

File: src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md
Comment (Youssef1313) [inline src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:6]: Shouldn't CA1837 be reused based on https://github.com/dotnet/runtime/issues/43257#issuecomment-713061245?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: I think this should use SymbolEqualityComparer.Default

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersFixer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersFixer.cs:54]: It may be better to add a third case, and make the default case throw.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:13]: nit: Include the other rules as well.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs:3]: nit: rename test file name.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:63]: I think `threadType` isn't required by all rules. I know it's unlikely that it doesn't exist, but just noting this anyway.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:15]: I'm not sure if it makes sense to extract an abstract class, and implement each rule separately?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs:112]: Should we have tests for something like:

```csharp
if (Process.GetCurrentProcess() is { Id: some_value })
{
    // do something
}
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:97]: nit: `SymbolEqualityComparer.Default.Equals(symbol1, symbol2)` will avoid the suppression.

**Review by Youssef1313 — COMMENTED** [review]
I think this should target .NET 6 branch (`6.0.1xx-preview3`).

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: > I think this should use SymbolEqualityComparer.Default

Why is that better than calling Equals?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:63]: It's not required, but it also exists everywhere that's relevant, so it didn't seem worthwhile to special-case.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:15]: I don't think it does.  From my perspective this is straightforward and less overhead.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: @stephentoub  I don't have full context to be honest. But that's what RS1024 analyzer recommends. The analyzer is currently disabled due to a crash bug, and will be re-enabled after it's fixed. If you enabled it, I believe you'll get a warning here.

Probably @sharwell or @mavasani have more context on why this is recommended.

File: src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md
Comment (stephentoub) [reply src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:6]: I don't think using the same diagnostic ID is a good idea here.  The target APIs were released in different versions of the libraries, and I think someone who's multitargeting will want to be able to suppress diagnostics for the APIs that don't exist in all of the platforms they're targeting while still getting diagnostics for those that do exist everywhere.  I'll need to rewatch the video to understand why we said that.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: Happy to change it if it's the right thing to do. But would like to understand. @sharwell ?

Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: `SymbolEqualityComparer` allows the author to be explicit on whether `string` and `string?` should be considered the same type (`SymbolEqualityComparer.Default`) or different (`SymbolEqualityComparer.IncludeNullability`). Code written before Nullable Reference Types used `Equals` but frequently failed to account for the difference here. The analyzer is a disambiguation aid for future maintainers who would otherwise be unsure whether the analyzer was written with NRT in mind.

Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: So I should use it to avoid a future analyzer yelling at me, but there's no functional difference in this case nor is there an ambiguity since none of the symbols in question are impacted by NRT. Correct?

Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: Yes, that is the action item here

Comment (jmarolf) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:94]: as @sharwell says, using the default equals is fine is nullability isn't a part of symbol identity in your case. The analyzer warns here because we want analyzer authors to be aware of this and make it explicit which behavior they indend.

**Comment by jmarolf** [comment]
@stephentoub as @Youssef1313 says, master is currently being used to service nuget packages. This PR should target https://github.com/dotnet/roslyn-analyzers/tree/release/6.0.1xx-preview3

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs
Comment (stephentoub) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs:112]: We could, but I don't think it's necessary.  There are lots of different ways to express the same thing, and all we really care about right now is the one that we've actually seen devs write.

**Comment by stephentoub** [comment]
Feedback addressed and rebased on release/6.0.1xx-preview3

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (sharwell) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:57]: 💡 In general, we use `context` as the name for all context parameters. The intentional aliasing prevents the invalid use of a context object from an enclosing scope.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs
Comment (sharwell) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs:62]: ❔ What reference assemblies do these first appear in?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs
Comment (Youssef1313) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentProcessIdTests.cs]: @stephentoub I didn't mean to necessarily support this, but at least to make sure that this doesn't cause the analyzer or fixer to crash.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs
Comment (stephentoub) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs:62]: CurrentManagedThreadId:
image

ProcessId:
image

ProcessPath:
Hasn't shipped yet, so .NET 6.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:57]: Okey dokey

**Comment by Youssef1313** [comment]
Test failure is unrelated and was fixed in https://github.com/dotnet/roslyn/pull/51244

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:57]:
```suggestion
            context.RegisterCompilationStartAction(context =>
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:60]:
```suggestion
                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEnvironment, out var environmentType) ||
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:61]:
```suggestion
                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsProcess, out var processType) ||
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:57]: Was this intentionally not renamed to context, or it was just missed?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:62]:
```suggestion
                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsProcessModule, out var processModuleType) ||
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:63]:
```suggestion
                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingThread, out var threadType))
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembers.cs:89]:
```suggestion
                context.RegisterOperationAction(context =>
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs:284]: The codefix *could* cleanup the now-unused usings. This is not necessary but probably good to have as a follow-up. Should I open an issue tracking this?

Comment (stephentoub) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseEnvironmentMembersTests.cs:284]: Do we do that with other analyzers? Removing usings seems orthogonal and a style preference,  not something we should be doing automatically just because we happened to remove the last use of something in that namespace. We could actually be creating more work and confusion for the developer.

**Comment by stephentoub** [comment]
@mavasani, any feedback before it's merged? Thanks.

**Review by mavasani — APPROVED** [review]
Looks great to me!

**Pull request merged**
